drivers/sensor/paw3212/paw3212.c

```diff
@@ -0,0 +1,602 @@
+		k_busy_wait(T_NCS_SCLK);
+	}
+
+	return err;
+}
+
+static int reg_read(struct paw3212_data *dev_data, u8_t reg, u8_t *buf)
+{
+	int err;
+
+	__ASSERT_NO_MSG((reg & SPI_WRITE_BIT) == 0);
+
+	err = spi_cs_ctrl(dev_data, true);
+	if (err) {
+		return err;
+	}
+
+	/* Write register address. */
+	const struct spi_buf tx_buf = {
+		.buf = &reg,
+		.len = 1
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	err = spi_write(dev_data->spi_dev, &spi_cfg, &tx);
+	if (err) {
+		LOG_ERR("Reg read failed on SPI write");
+		return err;
+	}
+
+	k_busy_wait(T_SRAD);
+
+	/* Read register value. */
+	struct spi_buf rx_buf = {
+		.buf = buf,
+		.len = 1,
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1,
+	};
+
+	err = spi_read(dev_data->spi_dev, &spi_cfg, &rx);
+	if (err) {
+		LOG_ERR("Reg read failed on SPI read");
+		return err;
+	}
+
+	err = spi_cs_ctrl(dev_data, false);
+	if (err) {
+		return err;
+	}
+
+	k_busy_wait(T_SRX);
+
+	return 0;
+}
+
+static int reg_write(struct paw3212_data *dev_data, u8_t reg, u8_t val)
+{
+	int err;
+
+	__ASSERT_NO_MSG((reg & SPI_WRITE_BIT) == 0);
+
+	err = spi_cs_ctrl(dev_data, true);
+	if (err) {
+		return err;
+	}
+
+	u8_t buf[] = {
+		SPI_WRITE_BIT | reg,
+		val
+	};
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = ARRAY_SIZE(buf)
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	err = spi_write(dev_data->spi_dev, &spi_cfg, &tx);
+	if (err) {
+		LOG_ERR("Reg write failed on SPI write");
+		return err;
+	}
+
+	k_busy_wait(T_SCLK_NCS_WR);
+
+	err = spi_cs_ctrl(dev_data, false);
+	if (err) {
+		return err;
+	}
+
+	k_busy_wait(T_SWX);
+
+	return 0;
+}
+
+static void irq_handler(struct device *gpiob, struct gpio_callback *cb,
+			u32_t pins)
+{
+	sensor_trigger_handler_t handler;
+
+	k_spinlock_key_t key = k_spin_lock(&paw3212_data.lock);
+	handler = paw3212_data.data_ready_handler;
+	k_spin_unlock(&paw3212_data.lock, key);
+
+	struct sensor_trigger trig = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_ALL,
+	};
+
+	handler(DEVICE_GET(paw3212), &trig);
+}
+
+static int paw3212_async_init_power_up(struct paw3212_data *dev_data)
+{
+	/* Reset sensor */
+
+	return reg_write(dev_data, PAW3212_REG_POWER_UP_RESET, 0x80);
+}
+
+static int paw3212_async_init_verify_id(struct paw3212_data *dev_data)
+{
+	int err;
+
+	u8_t product_id;
+	err = reg_read(dev_data, PAW3212_REG_PRODUCT_ID, &product_id);
+	if (err) {
+		LOG_ERR("Cannot obtain product ID");
+		return err;
+	}
+
+	LOG_INF("Product ID: 0x%" PRIx8, product_id);
+	if (product_id != PAW3212_PRODUCT_ID) {
+		LOG_ERR("Invalid product ID (0x%" PRIx8")!", product_id);
+		return -EIO;
+	}
+
+	return err;
+}
+
+static int paw3212_async_init_configure(struct paw3212_data *dev_data)
+{
+	int err;
+
+	err = reg_write(dev_data, PAW3212_REG_WRITE_PROTECT, PAW3212_WPMAGIC);
+	if (!err) {
+		u8_t mouse_option = 0;
+
+		if (IS_ENABLED(CONFIG_PAW3212_ORIENTATION_90)) {
```

Comment:
@jhn-nordic , I am wondering... in your code you swap x axis only. Considering the sensor rotation in the mouse chassis both axes should be swap. That would assume one of then is already swapped somehow.

I added the rotation as specified in datasheet and what I see is that one of the axes (y) is incorrectly swapped for some reason. I am wondering what can cause this. The other sensor does not show the same behavior.

Reply:
https://docs.microsoft.com/en-us/windows/desktop/gdi/window-coordinate-system

Coordinates 0,0 is in the upper left corner.